### PR TITLE
feat(py/compat-oai): add multimodal model support (image, TTS, STT)

### PR DESCRIPTION
**Pull request by yesudeep**
## Summary

Add image generation, text-to-speech, and speech-to-text capabilities to the Python `compat-oai` plugin, achieving feature parity with the JavaScript implementation. Update the `compat-oai-hello` sample to demonstrate all multimodal flows.

## Motivation

The JS `compat-oai` plugin supports image generation (DALL-E 3, GPT-Image-1), TTS (tts-1, tts-1-hd, gpt-4o-mini-tts), and STT (whisper-1, gpt-4o-transcribe, gpt-4o-mini-transcribe) models. The Python plugin was missing these capabilities. This change closes that gap.

### Plugin — New files
- **`models/image.py`** — `OpenAIImageModel` for image generation via `client.images.generate()`
- **`models/audio.py`** — `OpenAITTSModel` (via `client.audio.speech.create()`) and `OpenAISTTModel` (via `client.audio.transcriptions.create()`)
- **`models/utils.py`** — Shared helpers: `extract_text_from_parts()`, `extract_first_media_url()`, `parse_data_uri()`, and `data_uri_to_bytes()`

### Plugin — Modified files
- **`models/__init__.py`** — Export new model classes and registries (`SUPPORTED_IMAGE_MODELS`, `SUPPORTED_TTS_MODELS`, `SUPPORTED_STT_MODELS`)
- **`openai_plugin.py`** — Register, resolve, and list image/TTS/STT models using name-based pattern matching

### Sample — Updated files
- **`compat-oai-hello/src/main.py`** — Added three new flows:
  - `generate_image` — DALL-E image generation (returns base64 data URI)
  - `text_to_speech` — TTS with voice selection (returns base64 audio data URI)
  - `round_trip_tts_stt` — Self-contained Text → Speech → Text demo (generates audio via TTS, then transcribes it back via Whisper, fully testable from the Dev UI with just text input)
  - Added `_extract_media_url()` helper for extracting `MediaPart` data URIs from responses (since `response.text` returns empty for media-only responses)
- **`compat-oai-hello/README.md`** — Updated testing checklist with multimodal flows

### Tests
- **`image_model_test.py`** — 15 unit tests covering prompt extraction, param conversion, response conversion, model registry, and end-to-end model class
- **`audio_model_test.py`** — 19 unit tests covering text/media extraction, TTS/STT param conversion, response conversion, model registries, and model classes
- **`utils_test.py`** — Unit tests for shared utility functions

## Model Specs

| Type | Models | `supports.media` | `supports.output` | JS Parity |
|------|--------|-------------------|--------------------|-----------| 
| **Image** | dall-e-3, gpt-image-1 | `False` | `["media"]` | ✅ |
| **TTS** | tts-1, tts-1-hd, gpt-4o-mini-tts | `False` | `["media"]` | ✅ |
| **STT** | whisper-1, gpt-4o-transcribe, gpt-4o-mini-transcribe | `True` | `["text", "json"]` | ✅ |

## Design Decisions

- **MediaPart extraction**: `response.text` only extracts `TextPart`s, returning empty string for media-only responses (image, TTS). Flows and the shared `_extract_media_url()` helper iterate through `response.message.content` to find `MediaPart` data URIs.
- **Round-trip STT flow**: Instead of requiring raw base64 audio input (impractical to test in the Dev UI and fails with fake data), the `round_trip_tts_stt` flow generates audio via TTS first, then transcribes it — making it self-contained and testable with just a text input.

## Verification

- [x] `ruff check` — All checks passed
- [x] `pyrefly check` — 0 errors
- [x] `pytest` — All tests passed (34 new + 20 existing)
- [x] All pre-push hooks passed

catto
catto2

### CODE REVIEW

**Review by MengqinShen — APPROVED** [review]

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py [hunk 1]
```diff
@@ -78,23 +84,36 @@ def __init__(self, **openai_params: Any) -> None:  # noqa: ANN401
         """
         self._openai_params = openai_params
         self._openai_client = OpenAIClient(**openai_params)
+        self._async_client = AsyncOpenAI(**openai_params)
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py:209]: Good catch on the sync/async inconsistency. The pre-existing `OpenAIModel` using a sync client inside async methods is indeed a concern, but it's outside the scope of this PR. I've filed it as a follow-up. For the new multimodal models, they correctly use the `AsyncOpenAI` client throughout.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+        raise ValueError('No messages found in the request')
+
+    for part in request.messages[0].content:
+        if isinstance(part.root, MediaPart) and part.root.media:
+            media = part.root.media
+            content_type = str(media.content_type or '')
+            url = str(media.url or '')
+            if not content_type and url.startswith('data:'):
+                # Extract content type from data URI.
+                content_type = url[len('data:') : url.index(';')]
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Added a try/except ValueError to gracefully handle malformed data URIs without semicolons.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+    for key in ('temperature', 'maxOutputTokens', 'stopSequences', 'topK', 'topP'):
+        config.pop(key, None)
+
+    return {k: v for k, v in params.items() if v is not None}
+
+
+async def _to_tts_response(
+    response: HttpxBinaryResponseContent,
+    response_format: str = 'mp3',
+) -> GenerateResponse:
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Made `_to_tts_response` synchronous — `response.read()` on `HttpxBinaryResponseContent` is indeed a sync call. Updated the caller to remove the `await`.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+        # Transcription | TranscriptionVerbose | TranscriptionDiarized | str.
+        # Pyright sees the full overloaded union including AsyncStream.
+        # Extract text safely for all concrete return types.
+        if isinstance(result, str):
+            text = result
+        elif hasattr(result, 'text'):
+            text = result.text  # pyright: ignore[reportAttributeAccessIssue] -- stream=False excludes AsyncStream
+        else:
+            text = str(result)
+        return _to_stt_response(text)
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Moved all the text extraction logic into `_to_stt_response` which now handles the full union (`Transcription | str` + hasattr fallback). The `generate()` method now simply passes the raw SDK result through.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py [hunk 5]
```diff
@@ -172,6 +204,69 @@ def _create_model_action(self, name: str) -> Action:
...
+        stt_model = OpenAISTTModel(clean_name, self._async_client)
+        model_info = SUPPORTED_STT_MODELS.get(clean_name)
+        info_dict = model_info.model_dump(by_alias=True, exclude_none=True) if model_info else {}
+
+        return Action(
+            kind=ActionKind.MODEL,
+            name=name,
+            fn=stt_model.generate,
+            metadata={'model': info_dict},
+        )
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py]: Done. Consolidated all three methods into a single `_create_multimodal_action(name, model_class, supported_models)` helper. Much cleaner.

File: py/plugins/compat-oai/tests/audio_model_test.py [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+    def test_extracts_text(self) -> None:
+        """Verify text extraction from a simple request."""
+        request = GenerateRequest(
+            messages=[
+                Message(role=Role.USER, content=[Part(root=TextPart(text='Hello'))]),
+            ],
+        )
+        got = _extract_text(request)
+        if got != 'Hello':
+            pytest.fail(f'got {got!r}, want "Hello"')
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/tests/audio_model_test.py]: Done. Switched all assertions to use `assert` statements for consistency with the rest of the test suite. Also switched from `hasattr` checks to `isinstance(part, MediaPart)` / `isinstance(part, TextPart)` for better type narrowing.

File: py/plugins/compat-oai/tests/image_model_test.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+        """Verify text extraction from a simple single-message request."""
+        request = GenerateRequest(
+            messages=[
+                Message(role=Role.USER, content=[Part(root=TextPart(text='a sunset'))]),
+            ],
+        )
+        got = _extract_prompt_text(request)
+        want = 'a sunset'
+        if got != want:
+            pytest.fail(f'_extract_prompt_text() = {got!r}, want {want!r}')
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/tests/image_model_test.py]: Done. Same fix — switched to `assert` statements throughout. Tests also now use `isinstance(part, MediaPart)` for better type safety.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/__init__.py [hunk 8]
```diff
@@ -29,10 +39,16 @@
...
+    'SUPPORTED_STT_MODELS',
+    'SUPPORTED_TTS_MODELS',
+    'OpenAIImageModel',
     'OpenAIModel',
     'OpenAIModelHandler',
+    'OpenAISTTModel',
+    'OpenAITTSModel',
     'PluginSource',
     'get_default_model_info',
 ]
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/__init__.py:54]: Done. Sorted `__all__` alphabetically.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+        model_name: The TTS model name (e.g., 'tts-1').
+        request: The Genkit generate request.
+
+    Returns:
+        A dictionary of parameters for client.audio.speech.create().
+    """
+    text = _extract_text(request)
+    config = request.config or {}
+    if not isinstance(config, dict):
+        config = config.model_dump(exclude_none=True) if hasattr(config, 'model_dump') else {}
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Good catch. Fixed across all three files (`audio.py` TTS + STT, `image.py`). Now using `.copy()` on dict configs and `.model_dump()` on Pydantic models to avoid mutating the original request.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/image.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        model_name: The OpenAI model name (e.g., 'dall-e-3').
+        request: The Genkit generate request.
+
+    Returns:
+        A dictionary of parameters for client.images.generate().
+    """
+    prompt = _extract_prompt_text(request)
+    config = request.config or {}
+    if not isinstance(config, dict):
+        config = config.model_dump(exclude_none=True) if hasattr(config, 'model_dump') else {}
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/image.py]: Done. Same fix applied here in `image.py` — config dict is now copied before mutation.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py [hunk 11]
```diff
@@ -266,6 +323,66 @@ async def list_actions(self) -> list[ActionMetadata]:
...
+                            'supports': Supports(
+                                media=True,
+                                output=['text', 'json'],
+                                multiturn=False,
+                                system_role=False,
+                                tools=False,
+                            ).model_dump(by_alias=True, exclude_none=True),
+                        },
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py]: Done. Extracted a `_multimodal_action_metadata(name, supported_models, default_supports)` module-level helper that handles the registry lookup and fallback logic. The three elif blocks now each call this helper with their respective model registry and default Supports.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 12]
```diff
@@ -0,0 +1,487 @@
...
+        if isinstance(request.config, dict):
+            config = request.config.copy()
+        elif hasattr(request.config, 'model_dump'):
+            config = request.config.model_dump(exclude_none=True)
+
+    # Decode the base64 audio data from the data URI.
+    if ',' in media_url:
+        b64_data = media_url[media_url.index(',') + 1 :]
+    else:
+        b64_data = media_url
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Using `media_url.split(',', 1)` instead of `media_url.index(',')` for more robust data URI parsing.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 13]
```diff
@@ -0,0 +1,487 @@
...
+    }
+
+    # Optional: prompt for context.
+    prompt_text = None
+    for part in request.messages[0].content:
+        if hasattr(part.root, 'text') and part.root.text:
+            prompt_text = part.root.text
+            break
+    if prompt_text:
+        params['prompt'] = prompt_text
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Replaced the loop with a `next()` generator expression using `isinstance(part.root, TextPart)` for type-safe prompt extraction.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/image.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    """
+    if not request.messages:
+        raise ValueError('No messages found in the request')
+
+    first_msg = request.messages[0]
+    for part in first_msg.content:
+        if isinstance(part.root, TextPart) and part.root.text:
+            return str(part.root.text)
+
+    raise ValueError('No text content found in the first message')
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/image.py]: Done. Consolidated `_extract_prompt_text` in image.py to reuse `_extract_text` from audio.py. image.py now imports and re-exports it as `_extract_prompt_text` for backward compatibility.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py [hunk 15]
```diff
@@ -130,18 +180,31 @@ def get_model_info(self, name: str) -> dict[str, Any] | None:
...
+            # Use pattern matching on the name to determine the model type,
+            # following the same logic as the JS openai/index.ts resolver.
+            if 'gpt-image' in name or 'dall-e' in name:
+                return self._create_multimodal_action(name, OpenAIImageModel, SUPPORTED_IMAGE_MODELS)
+            elif 'tts' in name:
+                return self._create_multimodal_action(name, OpenAITTSModel, SUPPORTED_TTS_MODELS)
+            elif 'whisper' in name or 'transcribe' in name:
+                return self._create_multimodal_action(name, OpenAISTTModel, SUPPORTED_STT_MODELS)
+            else:
+                return self._create_model_action(name)
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py]: Done. Introduced a `_ModelType` enum and `_classify_model(name)` helper that centralizes all name-based pattern matching. Both `resolve()` and `list_actions()` now use this single function, along with `_MULTIMODAL_CONFIG` and `_DEFAULT_SUPPORTS` lookup tables.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py [hunk 16]
```diff
@@ -172,6 +235,34 @@ def _create_model_action(self, name: str) -> Action:
...
+        model_instance = model_class(clean_name, self._async_client)
+        model_info = supported_models.get(clean_name)
+        info_dict = model_info.model_dump(by_alias=True, exclude_none=True) if model_info else {}
+
+        return Action(
+            kind=ActionKind.MODEL,
+            name=name,
+            fn=model_instance.generate,
+            metadata={'model': info_dict},
+        )
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py:355]: Done. `_create_multimodal_action` now accepts a `model_type` parameter and uses `_DEFAULT_SUPPORTS[model_type]` to provide fallback metadata (label + supports) for dynamically discovered models not in the registry, consistent with `_multimodal_action_metadata`.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 17]
```diff
@@ -0,0 +1,489 @@
...
+
+    Raises:
+        ValueError: If no text content is found.
+    """
+    if not request.messages:
+        raise ValueError('No messages found in the request')
+
+    for part in request.messages[0].content:
+        if isinstance(part.root, TextPart) and part.root.text:
+            return str(part.root.text)
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Removed the redundant `str()` cast — `part.root.text` is already a `str` after the `isinstance(part.root, TextPart)` check.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 18]
```diff
@@ -0,0 +1,489 @@
...
+        ValueError: If no media content is found.
+    """
+    if not request.messages:
+        raise ValueError('No messages found in the request')
+
+    for part in request.messages[0].content:
+        if isinstance(part.root, MediaPart) and part.root.media:
+            media = part.root.media
+            content_type = str(media.content_type or '')
+            url = str(media.url or '')
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Simplified to `url = media.url` — it's a required `str` field, so the fallback and cast were unnecessary.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py [hunk 19]
```diff
@@ -78,23 +186,57 @@ def __init__(self, **openai_params: Any) -> None:  # noqa: ANN401
...
+        # Add STT models.
+        for name in SUPPORTED_STT_MODELS:
+            actions.append(
+                self._create_multimodal_action(
+                    open_ai_name(name),
+                    OpenAISTTModel,
+                    SUPPORTED_STT_MODELS,
+                    _ModelType.STT,
+                )
+            )
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py]: Done. Consolidated the three separate loops into a single loop over `_MULTIMODAL_CONFIG.items()`.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 20]
```diff
@@ -0,0 +1,489 @@
...
+            config = request.config.copy()
+        elif hasattr(request.config, 'model_dump'):
+            config = request.config.model_dump(exclude_none=True)
+
+    # Decode the base64 audio data from the data URI.
+    if ',' in media_url:
+        _, b64_data = media_url.split(',', 1)
+    else:
+        b64_data = media_url
+    audio_bytes = base64.b64decode(b64_data)
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Added explicit handling for data URIs (`data:` prefix), remote URLs (`http://`/`https://` which raise a clear ValueError), and raw base64 as fallback.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py [hunk 21]
```diff
@@ -130,17 +251,25 @@ def get_model_info(self, name: str) -> dict[str, Any] | None:
...
         """
+        if action_type == ActionKind.EMBEDDER:
+            return self._create_embedder_action(name)
+
         if action_type == ActionKind.MODEL:
+            model_type = _classify_model(name)
+            if model_type in _MULTIMODAL_CONFIG:
+                model_class, supported_models = _MULTIMODAL_CONFIG[model_type]
+                return self._create_multimodal_action(name, model_class, supported_models, model_type)
             return self._create_model_action(name)
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py:296]: Done. Added `if model_type == _ModelType.EMBEDDER: return None` guard in the MODEL branch to prevent embedder names from being incorrectly resolved as chat models.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 22]
```diff
@@ -0,0 +1,489 @@
...
+    }
+
+    prompt_text = next(
+        (
+            str(part.root.text)
+            for part in request.messages[0].content
+            if isinstance(part.root, TextPart) and part.root.text
+        ),
+        None,
+    )
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Extracted `_find_text(request) -> str | None` into utils.py. `_extract_text` now delegates to it and raises on None. `_to_stt_params` uses `_find_text` directly for optional prompt extraction.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 23]
```diff
@@ -0,0 +1,489 @@
...
+    custom_format = config.pop('response_format', None)
+
+    if custom_format:
+        params['response_format'] = custom_format
+    elif output_format == 'json':
+        params['response_format'] = 'json'
+    elif output_format == 'text':
+        params['response_format'] = 'text'
+    else:
+        params.setdefault('response_format', 'text')
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Simplified to the suggested pattern — single `response_format` variable with config pop, output.format check, and 'text' default.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/image.py [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+from __future__ import annotations
+
+from typing import Any
+
+from openai import AsyncOpenAI
+from openai.types.images_response import ImagesResponse
+
+from genkit.ai import ActionRunContext
+from genkit.core.typing import FinishReason
+from genkit.plugins.compat_oai.models.audio import _extract_text
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/image.py]: Done. Moved `_extract_text` and `_find_text` into `models/utils.py`. Both `audio.py` and `image.py` now import from utils, eliminating the cross-dependency between sibling model modules.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/utils.py [hunk 25]
```diff
@@ -32,6 +33,46 @@
...
+        The text content, or None if no text is found.
+    """
+    if not request.messages:
+        return None
+
+    for part in request.messages[0].content:
+        if isinstance(part.root, TextPart) and part.root.text:
+            return part.root.text
+
+    return None
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/utils.py]: Done. Replaced the loop with `next()` generator expression.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py [hunk 26]
```diff
@@ -242,21 +410,19 @@ async def embed_fn(request: EmbedRequest) -> EmbedResponse:
...
-                - name (str): The name of the action or model.
-                - kind (ActionKind): The type or category of the action.
-                - info (dict): The metadata dictionary describing the model configuration and properties.
-                - config_schema (type): The schema class used for validating the model's configuration.
+            list[ActionMetadata]: A list of ActionMetadata objects.
         """
-        actions = []
+        actions: list[ActionMetadata] = []
         models_ = self._openai_client.models.list()
         models: list[Model] = models_.data
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py]: Done. Switched `list_actions` to use `await self._async_client.models.list()` and updated the test to mock `_async_client` with `AsyncMock`.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/__init__.py [hunk 27]
```diff
@@ -28,11 +38,17 @@
...
+    'OpenAITTSModel',
     'PluginSource',
-    'get_default_model_info',
+    'SUPPORTED_EMBEDDING_MODELS',
+    'SUPPORTED_IMAGE_MODELS',
+    'SUPPORTED_OPENAI_COMPAT_MODELS',
+    'SUPPORTED_OPENAI_MODELS',
+    'SUPPORTED_STT_MODELS',
+    'SUPPORTED_TTS_MODELS',
 ]
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/__init__.py:54]: Done. Sorted `__all__` alphabetically — moved `get_default_model_info` after the uppercase names.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 28]
```diff
@@ -0,0 +1,458 @@
...
+    Raises:
+        ValueError: If no media content is found.
+    """
+    if not request.messages:
+        raise ValueError('No messages found in the request')
+
+    for part in request.messages[0].content:
+        if isinstance(part.root, MediaPart) and part.root.media:
+            media = part.root.media
+            content_type = str(media.content_type or '')
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Removed the redundant `str()` cast — `media.content_type or ''` already evaluates to `str`.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 29]
```diff
@@ -0,0 +1,458 @@
...
+            url = media.url
+            if not content_type and url.startswith('data:'):
+                # Extract content type from data URI.
+                try:
+                    content_type = url[len('data:') : url.index(';')]
+                except ValueError:
+                    pass
+            return url, content_type
+
+    raise ValueError('No media content found in the first message')
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Refactored to use `next()` for find-first, then process the found part separately.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/image.py [hunk 30]
```diff
@@ -0,0 +1,201 @@
...
+    content: list[Part] = []
+    for image in images:
+        if image.url:
+            url = image.url
+        elif image.b64_json:
+            url = f'data:image/png;base64,{image.b64_json}'
+        else:
+            continue
+
+        content.append(Part(root=MediaPart(media=Media(content_type='image/png', url=url))))
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/image.py]: Done. Simplified — compute `url` first from `image.url` or `b64_json`, then conditionally append.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py [hunk 31]
```diff
@@ -172,6 +303,43 @@ def _create_model_action(self, name: str) -> Action:
...
+                'label': f'OpenAI - {clean_name}',
+                'supports': default_supports.model_dump(by_alias=True, exclude_none=True) if default_supports else {},
+            }
+
+        return Action(
+            kind=ActionKind.MODEL,
+            name=name,
+            fn=model_instance.generate,
+            metadata={'model': info_dict},
+        )
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py:355]: Done. Extracted `_get_multimodal_info_dict()` helper that both `_multimodal_action_metadata` and `_create_multimodal_action` now call. Updated `_multimodal_action_metadata` signature to accept `model_type` instead of `default_supports`.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py [hunk 32]
```diff
@@ -130,17 +270,27 @@ def get_model_info(self, name: str) -> dict[str, Any] | None:
...
+
         Args:
             action_type: The kind of action to resolve.
             name: The namespaced name of the action to resolve.
 
         Returns:
             Action object if found, None otherwise.
         """
+        if action_type == ActionKind.EMBEDDER:
+            return self._create_embedder_action(name)
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py:287]: Done. Added symmetric guard — non-embedder names now return `None` when resolved as `ActionKind.EMBEDDER`, matching the existing guard in the MODEL branch.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 33]
```diff
@@ -0,0 +1,461 @@
...
+    # Determine file extension from content type for the filename.
+    ext_map = {
+        'audio/mpeg': 'mp3',
+        'audio/mp3': 'mp3',
+        'audio/wav': 'wav',
+        'audio/ogg': 'ogg',
+        'audio/flac': 'flac',
+        'audio/webm': 'webm',
+        'audio/mp4': 'mp4',
+    }
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Moved to module-level constant `_CONTENT_TYPE_TO_EXTENSION`.

File: py/GEMINI.md [hunk 34]
```diff
@@ -3079,3 +3079,261 @@ done
...
+  (e.g., `audio.py` and `image.py`) have identical helper functions, consolidate
+  into one and import from the other. Re-export with an alias if needed for
+  backward compatibility.
+* **Extract shared info dict builders.** When the same metadata serialization
+  logic (e.g., `model_info.model_dump()` with fallback) appears in both Action
+  creation and ActionMetadata creation, extract a single helper like
+  `_get_multimodal_info_dict(name, model_type, registry)` and call it from both.
+* **Re-assert `isinstance` after `next()` for type narrowing.** Type checkers
+  can't track narrowing inside generator expressions. After `next()`, re-assert
+  `isinstance(part.root, MediaPart)` locally so the checker can narrow:
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/GEMINI.md:3129]: Done. Added the code example and also documented the module-level constant hoisting pattern.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 35]
```diff
@@ -0,0 +1,462 @@
...
+
+    # Decode the base64 audio data from the data URI.
+    if media_url.startswith('data:'):
+        _, b64_data = media_url.split(',', 1)
+        audio_bytes = base64.b64decode(b64_data)
+    elif media_url.startswith(('http://', 'https://')):
+        raise ValueError(f'Remote audio URLs are not supported; provide a base64 data URI instead: {media_url[:50]}...')
+    else:
+        # Assume raw base64 string.
+        audio_bytes = base64.b64decode(media_url)
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Added try/except with descriptive ValueError for both data URI and raw base64 decoding paths.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 36]
```diff
@@ -0,0 +1,462 @@
...
+    # Optional STT-specific params.
+    for key in ('language', 'timestamp_granularities'):
+        if key in config:
+            params[key] = config.pop(key)
+
+    # Determine response format from output config.
+    response_format = config.pop('response_format', None)
+    if not response_format and request.output and request.output.format in ('json', 'text'):
+        response_format = request.output.format
+    params['response_format'] = response_format or 'text'
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py:295]: Done. Collapsed to single expression with `or` chain.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py [hunk 37]
```diff
@@ -53,6 +60,126 @@ def open_ai_name(name: str) -> str:
...
+        system_role=False,
+        tools=False,
+    ),
+}
+
+# Maps multimodal model types to their class and registry.
+_MULTIMODAL_CONFIG: dict[
+    _ModelType,
+    tuple[type[OpenAIImageModel | OpenAITTSModel | OpenAISTTModel], dict[str, ModelInfo]],
+] = {
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/openai_plugin.py]: Done. Created `_MultimodalModel` and `_MultimodalModelConfig` TypeAliases.

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 38]
```diff
@@ -0,0 +1,462 @@
...
+    assert isinstance(part_with_media.root, MediaPart)
+    media = part_with_media.root.media
+    content_type = media.content_type or ''
+    url = media.url
+    if not content_type and url.startswith('data:'):
+        # Extract content type from data URI.
+        try:
+            content_type = url[len('data:') : url.index(';')]
+        except ValueError:
+            pass
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Extracted data URI parsing into `utils.py` as `parse_data_uri_content_type()` and `decode_data_uri_bytes()`. Both `_extract_media()` and `_to_stt_params()` now use these utilities. Also extracted `extract_config_dict()` for the duplicated config-copying pattern. Added 37 exhaustive tests in `utils_test.py` covering all edge cases (malformed URIs, empty strings, remote URLs, charset params, etc.).

File: py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py [hunk 39]
```diff
@@ -0,0 +1,413 @@
...
+    for key in ('language', 'timestamp_granularities'):
+        if key in config:
+            params[key] = config.pop(key)
+
+    # Determine response format: config override > output format > default.
+    response_format = config.pop('response_format', None) or (
+        request.output.format if request.output and request.output.format in ('json', 'text') else None
+    )
+    params['response_format'] = response_format or 'text'
+
```
[the remark this answers is not in the document]
Comment (yesudeep) [reply py/plugins/compat-oai/src/genkit/plugins/compat_oai/models/audio.py]: Done. Separated the config pop from the output format fallback into distinct steps.
